qiime/beta_diversity.py

```diff
@@ -208,17 +208,23 @@ def single_file_beta(input_path, metrics, tree_path, output_dir,
                         sample_ids = otu_table.ids()
                         observation_ids = otu_table.ids(axis='observation')
                         for i in range(len(sample_ids)):
+                            samp_a = otu_table.data(sample_ids[rowidx])
```

Comment:
Fast Unifrac does not require a dense representation. It does cache
information on the tree, but it doesn't use the zeros. However, if enabling
fast unifrac to scale within qiime requires modification of pycogent, then
I agree it is not worth pursuing.

Do you know why one_sample_unweighted_unifrac is being used instead of
dist_unweighted_unifrac? My impression was that the latter would work and
what I was expecting the metric being used to be. It is also fast unifrac

On Tue, Jun 9, 2015 at 3:39 PM, josenavas <notifications@github.com> wrote:

> In qiime/beta_diversity.py
> <https://github.com/biocore/qiime/pull/2040#discussion_r32067547>:
>
> > @@ -208,17 +208,23 @@ def single_file_beta(input_path, metrics, tree_path, output_dir,
> >                          sample_ids = otu_table.ids()
> >                          observation_ids = otu_table.ids(axis='observation')
> >                          for i in range(len(sample_ids)):
> > +                            samp_a = otu_table.data(sample_ids[rowidx])
>
> @wasade <https://github.com/wasade> after looking at the code with
> @ElDeveloper <https://github.com/ElDeveloper>, it turns out these lines
> inside the AttributeError will never be executed for the unifrac
> distances, as line 204 do not raise any error.
>
> IIRC, the current implementation of Unifrac in cogent is FastUnifrac.
> FastUnifrac is able to compute unifrac fast because it pre-computes a lot
> of data structures. that is why for Unifrac these lines never get executed.
>
> One way of doing this will be to create a for loop and give the code an
> OTU table with 2 samples each time, and then just construct the distance
> matrix ourselves in here. On one hand, this will kill the "Fast" part of
> Unifrac, since it will recompute those structures for each pair of samples.
> On the other hand, we will be able to aggressively parallelize this, e.g.
> creating a job for each pair of samples; maximizing the parallelization
> during those pre-computations.
>
> Other options require more though on how to parallelize this, instead of
> using the naïve per-row parallelization.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/qiime/pull/2040/files#r32067547>.
>